packages/trader/src/App/Components/Elements/market-countdown-timer.tsx

```diff
@@ -0,0 +1,224 @@
+                    if (!api_response.api_initial_load_error) {
+                        const returned_symbol = getSymbol(target_symbol, api_response.trading_times);
+                        const open = returned_symbol?.times.open as string[];
+                        const close = returned_symbol?.times.close as string[];
+                        const is_closed_all_day = open?.length === 1 && open[0] === '--' && close[0] === '--';
+                        if (is_closed_all_day) {
+                            // check tomorrow trading times
+                            return whenMarketOpens(days_offset + 1, target_symbol);
+                        }
+                        const date_str = target_date.toISOString().substring(0, 11);
+                        const getUTCDate = (hour: string) => new Date(`${date_str}${hour}Z`);
+                        for (let i = 0; i < open?.length; i++) {
+                            const diff = +getUTCDate(open[i]) - Date.now();
+                            if (diff > 0) {
+                                remaining_time_to_open = +getUTCDate(open[i]);
+                                if (isMounted() && target_symbol === symbol) {
+                                    return setWhenMarketOpens({
+                                        days_offset,
+                                        opening_time: open[i],
+                                        remaining_time_to_open,
+                                    });
+                                }
+                            }
+                        }
+                        whenMarketOpens(days_offset + 1, target_symbol);
+                    }
+                };
+
+                whenMarketOpens(0, symbol);
+            }
+            // eslint-disable-next-line react-hooks/exhaustive-deps
+        }, [symbol]);
+
+        React.useEffect(() => {
+            let timer: ReturnType<typeof setTimeout>;
+            if (when_market_opens?.remaining_time_to_open) {
+                timer = setTimeout(() => {
+                    setTimeLeft(calculateTimeLeft(when_market_opens.remaining_time_to_open));
+                    if (+new Date(when_market_opens.remaining_time_to_open) - +new Date() < 1000) {
+                        setLoading(true);
+                        if (is_main_page) onMarketOpen(false);
+                    }
+                }, 1000);
+            }
+            return () => {
+                if (timer) {
+                    clearTimeout(timer);
+                }
+            };
+        }, [time_left, when_market_opens, onMarketOpen, is_main_page]);
+
+        React.useEffect(() => {
+            if (!is_loading) setIsTimerLoading(false);
+        }, [is_loading, setIsTimerLoading]);
+
+        let timer_components = '';
+
+        if (Object.keys(time_left).length) {
+            const hours = ((time_left.days ?? 0) * 24 + (time_left.hours ?? 0)).toString().padStart(2, '0');
```

Comment:
does it make sense to get rid of undefined type like this `time_left.days ?? 0` @maryia-deriv

Reply:
@henry-deriv  as I can see, `time_left` is set to a value returned from `calculateTimeLeft()` which is either an empty object where all properties will be `undefined`, or an object containing `{days: 0, hours: 11, minutes: 54, seconds: 39}` . And here we are protected by the condition where you check that `time_left` has properties but we can also add `'days' in time_left && 'hours' in time_left && 'minutes' in time_left && 'seconds' in time_left` to the `if (Object.keys(time_left).length) {` condition, and see if TS is satisfied. If not, you will have to use the current way of adding `?? 0` or wrap it into `Number` like so: `Number(time_left.days) `

Reply:
i wrapped them in Number. property check didnt worrk.